### ISSUE TITLE
Support hierarchical project preferences

### ISSUE THREAD

**Issue by laeubi** [eclipse-platform/eclipse.platform#89]

**Comment by HannesWell** [comment]
IIRC @merks started to prototype something related in the context of Oomph a (long) while ago?

**Comment by merks** [comment]
Yes, long ago I started implementing something that would automatically copy/propagate project-specific preferences between projects where one could specify which project is the "prototype" from which the preferences should be copied.  But I never completed this work such that it's actually in use by any project...

**Comment by laeubi** [comment]
Are there some comment on how this works? It has been a while since I wrote this down, but now I think my initial proposal could even be adjusted by instead of

> Given there is a file .settings/org.eclipse.jdt.core.prefs nothing changes

to simply delegate e.g. overwrite anything from the lower levels so we have:

1. workspace values
2. parent project values 
3. project specific values

and the highest value wins.

My goal here would clearly be that one **do not** need to copy anything ... so once we might be able to have what @mickaelistria  likes from intelij/vscode that there is only one single folder in the workspace storing all ide specific configurations.

**Comment by laeubi** [comment]
> Are we talking here about `parent` from IFileSystem perspective or about IContainer?

I think the "parent" of a project is always the workspace so this seems not useful,  not sure how `IFileSystem` handles this, but for a local filesystem it should match the expectation of the user that sees a tree-hierarchy in its file browser.

> Is it supposed to work with Remote File Systems?

Why not? But I'm wondering if there are really people using a rmfs for their project to work on entirely?

**Comment by ruspl-afed** [comment]
So, for each IProject it will go up from its "Location URI" through the local file system, looking for `.eclipse` folder until the very root of the drive, and then will try to look into workspace? Or what?

**Comment by laeubi** [comment]
> So, for each IProject it will go up from its "Location URI" through the local file system, looking for `.eclipse` folder until the very root of the drive, and then will try to look into workspace? Or what?

yes, this is for example how maven works with the `.mvn` folder, I think one can have one exception when the project is inside the workspace folder then stop there.

**Comment by hd42** [comment]
I believe that Eclipse (at least with m2e) already has a notion of hierarchical project structures, which might be used to know how far to go "up", because when working with Git, the projects might not reside inside the workspace folder (for example /git/my.project/base/plugins/my.plugin vs. ~/workspaces/MyProjectWorkspace/), because workspace settings are not committed to git.

**Comment by HannesWell** [comment]
While working on https://github.com/eclipse-equinox/equinox/pull/446, I also thought about this and how to implement this.

In general I think this a good approach, but it should only be implemented in the _Project_-preference scope so that the project-scope in itself is hierarchical, while being the most specific scope of the project, instance/workspace, configuration/installation and hopefully soon user scope (plus the default scopes) hierarchy.
I think this makes sense since the proposed project preferences are tied to the location of the project in the file-system, while the location of the `project`, `workspace` and `installation` (and `user`) are in general absolutely independent and can all have their own location.

If a value is search in the project scope, the closest preference file should be searched first and if the key is not found the further parents should be searched until a value for a key is found.

From an UI point of view, the shared preferences stored in an `.eclipse` folders could be modified like the project specific preferences that are stored inside the project-folder, simply by opening the `Properties` of the `.eclipse` folder via the (context) menu, just like it is done project specific preference now.
There could then also be a clean up to remove preference entries stored in a project that are duplicates from a parent directory preference (not considering from other scopes).

> To mitigate this and given that most projects have some hierarchical nature (e.g the usual `<projectroot>/plugins/<myproject>`) I'd like to propose the following enhancement (using `org.eclipse.jdt.core.prefs` as an example):
> 
>     * Given there is a file `.settings/org.eclipse.jdt.core.prefs` nothing changes

I would propose that nothing changes only if the requested key is not present in the existing preference file. If the requested key is missing it should be search in the parent directories `.eclipse/.settings/*.prefs`.

> 
>     * Given there is no file `.settings/org.eclipse.jdt.core.prefs` then first a file `.eclipse/org.eclipse.jdt.core.prefs` is searched and if found that is used, if that file is not found, the parent folder is checked for this until the root is reached
> 

I don't think we should have `.eclipse` as an alternative to the `.settings` folder within a project.
Furthermore the current implementation in `EclipsePreferences` assume that all `*.prefs` files are located in a `.settings` folder within the scope location, be it the project's location, the `instance` location in `.metadata\.plugins\org.eclipse.core.runtime\.settings` (i.e. the state location of the `o.e.core.runtime` in the current runtime instance or the `configuration` location.
In order to keep this consistent I think it an `.eclipse` folder in a parent directory should be considered as parent (within the `project` scope), which has the effect that the preference files are saved in `.eclipse/.settings/*.prefs`.
Only if we think that there will never be a reason to have other files/folders stored there we could think about changing `EclipsePreferences` to look in the preference scope directory directly.
Thinking about it, at least for projects, it could be an interesting to place all Eclipse specific files in the `.eclipse` folder, like the `.project`, `.classpath`, `META-INF/MANIFEST.MF`, `build.properties`, etc. But that's another story.

Furthermore by default only the parent directory should be considered in order to prevent potential confused if https://github.com/eclipse-equinox/equinox/pull/446 is implemented as suggested and a workspace is located within the user home plus to improve the performance (if performance is a problem depends on the exact implementation). Nevertheless it should be possible to configure the number of considered parent directories easily respectively to configure that the parent 


>     * If still no file is found `<workspace-root>/.eclipse/<project-name>/org.eclipse.jdt.core.prefs` is searched as a last resort

I think the workspace (scope) should not be considered here. This should only be done as part of the project scope. Of course it might happen that coincidentally the workspace root is equal to a parent directory containing a `.eclipse` preference folder.

**Comment by HannesWell** [comment]
Last year at OCX/EclipseCon 2024 I had a discussion with @stephan-herrmann about this (I'm a bit late on writing this down) and his suggestion was to establish the preference hierarchy respectively the parent-child relation by referencing the parent project by it's name. Then it could be located anywhere, e.g. colocated with other projects or in a `.eclipse/.settings` folder in the root of the repository or wherever it's suitable for a user.
The parent would then only be have to defined explicitly in the settings of a project. As I think hierarchical project preferences only make sense with a project, the `org.eclipse.core.resources.prefs` could fit for that.

**Comment by laeubi** [comment]
> The parent would then only be have to defined explicitly in the settings of a project. As I think hierarchical project preferences only make sense with a project, the org.eclipse.core.resources.prefs could fit for that.

One goal of hierarchical project preferences is to get rid of these `.settings` folders in most cases... having a setting to define where the settings are... this already works using links. This is all much to clumsy to setup and also "project name" the would require to check in `.project` files what often also is something one likes to avoid. Not talking about what happens if people using "settings" as this project name and you have 5 of them in different project and getting whatever strange effects and even others if people do not import the project at all.

So why we always need an "Eclipse way" instead of adopting what others use and work? e.g. `.mvn` in maven or `.vscode` and so on?

**Comment by HannesWell** [comment]
> One goal of hierarchical project preferences is to get rid of these `.settings` folders in most cases...

We could record the preference parent also in the `.project` file.

> So why we always need an "Eclipse way" instead of adopting what others use and work? e.g. `.mvn` in maven or `.vscode` and so on?

As written in my previous comment, you could then create the 'preference-parent' project used by all projects in the repo in a folder named `.eclipse` and make sure that that it contains a .project file where the project name is quite-unique, e.g. `eclipse-pde-settings`. The project name and folder name don't have to match. It would then just be an ordinary project without magic.
If you then really want to get rid of the `.settings` folder within that project we could also check if the location of the preferences in a project could become customizable (e.g. again in the .project file). Currently that's hard-coded (AFAICT) in:
- https://github.com/eclipse-equinox/equinox/blob/6ff2397f450a4ae927870981f06dbfa88b2e64ff/bundles/org.eclipse.equinox.preferences/src/org/eclipse/core/internal/preferences/EclipsePreferences.java#L47

Having dedicated preference-parent project also simplifies the UI as we can just re-use the existing facilities for modifying a project. If we just rely on file-path convenient, where in the UI would I modify the parent?

When importing projects we could also make sure that the preference parent is always imported as the .projects files are usually read there too. And if the parent is not in the selected folder, the importer could also be smart and search in the parents. That is done much more seldomly and therefore not such a big problem performance wise. 

I understand it's not easy to find a good solution, that's on hand easy and convenient to use but also fits in the existing model. But the latter should IMO also not be neglected because multiple different concepts also makes it harder to understand.

**Comment by laeubi** [comment]
> We could record the preference parent also in the .project file.

As mentioned, people often prefer to have only minimal "ide files" in the VCS so same problem here... beside that you need to configure it again for every project (and don't forget).

Apart form that, configure it per project is really od as it then obviously does not need to be "hierarchical" in any ways it can even be different or many of them, not even talking about what happens if they are cyclic ...

> As written in my previous comment, you could then create the 'preference-parent' project used by all projects in the repo in a folder named .eclipse and make sure that that it contains a .project file where the project name is quite-unique, e.g. eclipse-pde-settings. The project name and folder name don't have to match. It would then just be an ordinary project without magic.

The point of "hierarchical preferences" is they "work like magic" and I do not need to impose any restrictions (e.g. please name it unique world wide).

> I understand it's not easy to find a good solution, that's on hand easy and convenient to use but also fits in the existing model.

I don't think it fits in any way see above it just feels like a attempt to make an easy thing complex while we all try to reduce complexity. It would also introduce a somehow bad cycle where one preference need to be read once before other preferences can be read and soon someone will ask if we can not have "hierarchical preference for configuring the hierarchical preference"

> But the latter should IMO also not be neglected because multiple different concepts also makes it harder to understand.

This concept is used all over the place with different tools and if it is "hard to understand" I really wonder how people solve issues that are really hard... not talking about that a repository is usually set up once and used by many,

**Comment by mickaelistria** [comment]
> The parent would then only be have to defined explicitly in the settings of a project. 

Why doing it statically/explicitly when it can be relatively easily inferred dynamically from the project state?
